### ISSUE TITLE
Type checking error with functions that take BoaTypeVar

### ISSUE THREAD

**Issue by psybers** [boalang/compiler#17]
``` sh
s: stack of int;
push(s, s);
```

This code should give an error, as push would expect a stack and an int as parameters, but is given two stacks.

The compiler gives no error, instead leading to a javac compilation error:

```
Test.java:152: error: method push in class Stack<E> cannot be applied to given types;
                ___s.push(___s);
                    ^
  required: Long
  found: Stack<Long>
  reason: actual argument Stack<Long> cannot be converted to Long by method invocation conversion
  where E is a type-variable:
    E extends Object declared in class Stack
```
